src/main/java/edu/ucla/library/services/utility/PropsWriter.java

```diff
@@ -0,0 +1,165 @@
+        loadProps(args[0]);
+        readWrite(args[1], args[2]);
+    }
+
+    /**
+     * Public method for testing access or plugging into other classes.
+     *
+     * @param aDefaults default values fed to PROPS
+     * @param aTemplate template for the output file
+     * @param aOutput properties file built by app from aDefaults and aTemplate
+    */
+    public static void buildProperties(final String aDefaults,
+        final String aTemplate, final String aOutput) {
+        verifyArgs(new String[] {aDefaults, aTemplate, aOutput});
+        loadProps(aDefaults);
+        readWrite(aTemplate, aOutput);
+    }
+
+    /**
+     * Check that parameters are valid.
+     *
+     * @param aArray array of file names
+    */
+    public static void verifyArgs(final String[] aArray) {
+        if (aArray.length != ARGS_COUNT) {
+            System.err.println(
+                "usage: PropsWriter defaults template output");
+            System.exit(Constants.USAGE_ERROR);
+        }
+      // check that default props and template files exist/are readable
+        checkExists(aArray[0]);
+        checkReadable(aArray[0]);
+        checkExists(aArray[1]);
+        checkReadable(aArray[1]);
+    }
+
+    /**
+     * Verify a file exists.
+     *
+     * @param aFileName pathed name of file to test
+    */
+    public static void checkExists(final String aFileName) {
+        if (!Files.exists(FileSystems.getDefault().getPath(aFileName))) {
+            System.err.println("File must exist: " + aFileName);
+            System.exit(Constants.DOESNT_EXIST_ERROR);
+        }
+    }
+
+    /**
+     * Verify a file is readable.
+     *
+     * @param aFileName pathed name of file to test
+    */
+    public static void checkReadable(final String aFileName) {
+        if (!Files.isReadable(FileSystems.getDefault().getPath(aFileName))) {
+            System.err.println("File must be readable: " + aFileName);
+            System.exit(Constants.NOT_READABLE_ERROR);
+        }
+    }
+
+    /**
+     * Load default properties into PROPS.
+     *
+     * @param aFileName pathed name of default props file
+    */
+    public static void loadProps(final String aFileName) {
+        try {
+            PROPS.load(new FileInputStream(new File(aFileName)));
+        } catch (FileNotFoundException e) {
+            System.err.println("Props file must exist: "
+                + e.getMessage());
+            System.exit(Constants.DOESNT_EXIST_ERROR);
+        } catch (IOException details) {
+            System.err.println("I/O error reading props file: "
+                + details.getMessage());
+            System.exit(Constants.IO_ERROR);
+        }
+    }
+
+    /**
+     * Splice default properties in template and write to output file.
+     *
+     * @param aInput pathed name of template file
+     * @param aOutput pathed name of final properties file
+    */
+    public static void readWrite(final String aInput, final String aOutput) {
```

Comment:
One of the things that the original Python script is doing is replacing any properties from the default/tmpl files with ENV variables so that we have a way to override values from the properties files at runtime. Cf. for an example https://github.com/UCLALibrary/fester/blob/main/src/main/docker/docker-entrypoint.sh#L24

I don't think I'm seeing the same functionality here (unless I'm missing something). You can pass in ENVs to tests through the surefire plugin config in the POM. And it would be good to have a test that confirms the swap works.

Reply:
Added logic to check for ENV values, and use default props if not found